### PR TITLE
Updated Software Engineering/Permissioned Guard

### PR DESCRIPTION
**Pull request by nemild**
Changes:
- Added more context
- Added two examples (registry, delegate call)

### CODE REVIEW

File: smart-contracts.md [hunk 1]
```diff
@@ -356,11 +356,66 @@ When deploying on the main chain, one might want to the use the optimizer. This
...
+It's also critical to have a secure way for parties to upgrade the code - and code changes may be approved by a single trusted party, a group of members, or a vote of the full set of stakeholders. You may often combine a permissioned guard with a contract pause or lock, ensuring that activity does not occur while changes are being made - or after a contract is upgraded.
+
+**Example 1: Use a registry contract to store latest version of a contract**
+
+```
+contract SomeRegister {
+    address backendContract;
+    address[] previousBackends;
+
+    function changeBackend(address newBackend) {
```
Comment (simondlr) [inline smart-contracts.md]: Please test the code in browser-solidity for obvious syntax issues. requires returns bool here. Also. I would additionally change this example to include an owner. Anyone can change the backend here. Only the owner should be able to.

File: smart-contracts.md [hunk 2]
```diff
@@ -356,11 +356,66 @@ When deploying on the main chain, one might want to the use the optimizer. This
...
+Code will need to be changed if errors are discovered or if improvements need to be made - and there are various techniques to do this. The simplest is to have a registry contract that holds the address of the latest contract. A more seamless approach for contract users is to have a contract that forwards calls and data onto the latest version of the contract.
 
-A simple switch can also employed that allows the contracts to lock themselves, instead of having to worry about a fix. Ie, lock first, then upgrade.
+Whatever the technique, it's important to have modularization and good separation between components (data, logic) - so that code changes do not break functionality, orphan data, or require substantial costs to port.
+
+It's also critical to have a secure way for parties to upgrade the code - and code changes may be approved by a single trusted party, a group of members, or a vote of the full set of stakeholders. You may often combine a permissioned guard with a contract pause or lock, ensuring that activity does not occur while changes are being made - or after a contract is upgraded.
+
+**Example 1: Use a registry contract to store latest version of a contract**
+
```
Comment (simondlr) [inline smart-contracts.md]: Perhaps add here something like: "In this example, the calls aren't automatically forwarded. The users should fetch the current "head" contract from here, and interact with it separately." ?

File: smart-contracts.md [hunk 3]
```diff
@@ -356,11 +356,66 @@ When deploying on the main chain, one might want to the use the optimizer. This
...
+	}
+	_
+    }
+
+    function Relay(address initAddr) {
+        currentVersion = initAddr;
+        owner = msg.sender; // this owner may be another contract with multisig, not a single contract owner
+    }
+
+    function changeContract(newVersion)
```
Comment (simondlr) [inline smart-contracts.md]: Another syntax issue. address newVersion required. :)

**Comment by nemild** [comment]
Thanks, fixed those errors/issues!

**Comment by simondlr** [comment]
For some reason the indentation still isn't showing properly?

**Comment by nemild** [comment]
Had some issues with tabs vs spaces, hopefully this looks good now.

**Pull request merged**